src/layer/decorators.js

```diff
@@ -0,0 +1,143 @@
+    this.canvas = dom.createElement("canvas");
+    this.renderer = renderer;
+    this.pixelRatio = 1;
+    this.maxHeight = renderer.layerConfig.maxHeight;
+    this.lineHeight = renderer.layerConfig.lineHeight;
+    this.canvasHeight = parent.parent.scrollHeight;
+    this.heightRatio = this.canvasHeight / this.maxHeight;
+    this.canvasWidth = parent.width;
+    this.minDecorationHeight = (2 * this.pixelRatio) | 0;
+    this.halfMinDecorationHeight = (this.minDecorationHeight / 2) | 0;
+
+    this.canvas.width = this.canvasWidth;
+    this.canvas.height = this.canvasHeight;
+    this.canvas.style.top = 0  + "px";
+    this.canvas.style.right = 0 + "px";
+    this.canvas.style.zIndex = 7 + "px";
+    this.canvas.style.position = "absolute";
+    this.colors = {};
+    this.colors.dark = {
+        "error": "rgba(255, 18, 18, 1)",
+        "warning": "rgba(18, 136, 18, 1)",
+        "info": "rgba(18, 18, 136, 1)"
+    };
+
+    this.colors.light = {
+        "error": "rgb(255,51,51)",
+        "warning": "rgb(32,133,72)",
+        "info": "rgb(35,68,138)"
+    };
+
+    parent.element.appendChild(this.canvas);
+
+};
+
+(function () {
+    oop.implement(this, EventEmitter);
+
+    this.$updateDecorators = function (config) {
+        var colors = (this.renderer.theme.isDark === true) ? this.colors.dark : this.colors.light;
+        if (config) {
+            this.maxHeight = config.maxHeight;
+            this.lineHeight = config.lineHeight;
+            this.canvasHeight = config.height;
+            var allLineHeight = (config.lastRow + 1) * this.lineHeight;
+            if (allLineHeight < this.canvasHeight) {
+                this.heightRatio = 1;
+            } else {
+                this.heightRatio = this.canvasHeight / this.maxHeight;
+            }
+        }
+        var ctx = this.canvas.getContext("2d");
+
+        function compare(a, b) {
+            if (a.typeNumber < b.typeNumber) return -1;
+            if (a.typeNumber > b.typeNumber) return 1;
+            return 0;
+        }
+
+        var annotations = this.renderer.session.$annotations;
+        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
+        if (annotations) {
+            annotations.forEach(function (item) {
+                switch (item.type) {
+                    case "info":
+                        item.typeNumber = 1;
+                        break;
+                    case "warning":
+                        item.typeNumber = 2;
+                        break;
+                    case "error":
+                        item.typeNumber = 3;
+                        break;
+                }
+            });
+            annotations = annotations.sort(compare);
+            var foldData = this.renderer.session.$foldData;
+
+            for (let i = 0; i < annotations.length; i++) {
+                let currentRow = annotations[i].row;
+                let compensateFold = this.compensateFoldRows(currentRow, foldData);
```

Comment:
I'd rename currentRow to just row to avoid confusion with current cursor position